### PR TITLE
FullSearchMPI tuning strategy and WrapMPI to handle mpi calls inside of AutoPas

### PR DESCRIPTION
**Pull request by wothie**
# Description

- FullSearchMPI works similar to FullSearch, but splits the search space between all ranks and uses MPI calls to determine the optimal configuration.
- AutoPasConfigurationCommunicator was thus added as a helper for serialization and sending of configurations.
- WrapMPI is used internally to provide non-mpi versions of all necessary mpi calls in case MPI cannot be used. Otherwise it simply wraps the respective MPI calls without additional computation.
- The MPI communicator for AutoPas can now be set externally through AutoPas::setMPICommunicator
- A new testing folder has been added with tests for MPI methods. Currently only a test suite for FullSearchMPI exists

# Resolved Issue
 - None, but works on issue #410 

# How Has This Been Tested?

- [x] sph-main-mpi with AUTOPAS_MPI=ON/OFF and different numbers of processes (1, 2, or 4)
- [x] runMPITests with AUTOPAS_MPI=ON/OFF and different numbers of processes (1, 2, or 4)
- [x] Tested with both OpenMPI and MPICH
- [ ] Tested on cluster

### CODE REVIEW

File: src/autopas/utils/WrapMPI.h [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+  if (not initialized) {
+    result1 = MPI_Init(argc, argv);
+  }
+  result2 = MPI_Comm_dup(MPI_COMM_WORLD, &AUTOPAS_MPI_COMM_WORLD);
+  if (result1 == MPI_SUCCESS) {
+    return result2;
+  } else {
+    return result1;
+  }
+}
```
Comment (SteffenSeckler) [inline src/autopas/utils/WrapMPI.h:297]: move the communicator to AutoPas.h as setComm().

File: src/autopas/utils/WrapMPI.h [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+
+inline int AutoPas_MPI_Finalize() {
+  int finalized;
+  int result1 = MPI_SUCCESS;
+  int result2 = MPI_SUCCESS;
+  MPI_Finalized(&finalized);
+
+  if (not finalized) {
+    result1 = MPI_Comm_free(&AUTOPAS_MPI_COMM_WORLD);
```
Comment (SteffenSeckler) [inline src/autopas/utils/WrapMPI.h:306]: move to new AutoPas.finalize()

File: src/autopas/AutoPas.h [hunk 3]
```diff
@@ -75,13 +76,14 @@ class AutoPas {
         _relativeOptimumRange(1.2),
         _maxTuningPhasesWithoutTest(5),
         _acquisitionFunctionOption(AcquisitionFunctionOption::lowerConfidenceBound),
-        _tuningStrategyOption(TuningStrategyOption::fullSearch),
+        _tuningStrategyOption(TuningStrategyOption::fullSearchMPI),
```
Comment (SteffenSeckler) [inline src/autopas/AutoPas.h:79]: undo

File: src/autopas/AutoPas.h [hunk 4]
```diff
@@ -101,17 +102,27 @@ class AutoPas {
...
+            _maxTuningPhasesWithoutTest, _acquisitionFunctionOption, _autopasMPICommunicator)),
         _selectorStrategy, _tuningInterval, _numSamples);
     _logicHandler =
         std::make_unique<autopas::LogicHandler<Particle, ParticleCell>>(*(_autoTuner.get()), _verletRebuildFrequency);
   }
 
+  /**
+   * Free the AutoPas MPI communicator.
+   * DO call if: No MPI_Comm was not externally provided or is not freed externally
+   * Do NOT call if: Externally provided MPI_Comm is freed externally
```
Comment (SteffenSeckler) [inline src/autopas/AutoPas.h:122]: this is NOT a good idea!
this should always be called consistently.
as you are doing an `AutoPas_MPI_Comm_dup` it should be safe to free it, right?

File: src/autopas/AutoPas.h [hunk 5]
```diff
@@ -553,6 +564,23 @@ class AutoPas {
     _tuningStrategyOption = tuningStrategyOption;
   }
 
+// Only define the interface for the MPI communicator if AUTOPAS_MPI=ON
+// The internal implementation will use _autopasMPICommunicator with WrapMPI regardless of AUTOPAS_MPI
+#if defined(AUTOPAS_MPI)
+  /**
+   * Setter for the MPI communicator that AutoPas uses for potential MPI calls.
+   * If not set, MPI_COMM_WOLRD will be used.
```
Comment (SteffenSeckler) [inline src/autopas/AutoPas.h:572]:
```suggestion
   * If not set, MPI_COMM_WORLD will be used.
```

File: tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/FullSearchMPITest.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      {autopas::DataLayoutOption::soa, autopas::DataLayoutOption::aos}, {autopas::Newton3Option::enabled});
+
+  EXPECT_THROW(fullSearchMPI.removeN3Option(autopas::Newton3Option::enabled),
+               autopas::utils::ExceptionHandler::AutoPasException);
+}
+
+TEST_F(FullSearchMPITest, testGlobalOptimumAndReset) {
+  int rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+// manually set every rank to 0, because we cannot use WrapMPI here
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/FullSearchMPITest.cpp:50]: please add a short description that makes the idea of the test easily understandable.

File: src/autopas/utils/WrapMPI.h [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+
+/**
+ * Dummy for MPI_Datatype.
+ * initialize values to the size of the respective type in bytes.
+ */
+enum AutoPas_MPI_Datatype {
+  AUTOPAS_MPI_BYTE = 1,
+  AUTOPAS_MPI_CXX_BOOL = sizeof(bool),
+  AUTOPAS_MPI_INT = sizeof(int),
+  AUTOPAS_MPI_LONG_INT = sizeof(long) + sizeof(int),
```
Comment (SteffenSeckler) [inline src/autopas/utils/WrapMPI.h:86]: this is not necessarily true, see here:
https://godbolt.org/z/n5KCur

Best is, to remove this data type, it's anyways not used...

File: src/autopas/utils/WrapMPI.h [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+/** Dummy for MPI_STATUS_IGNORE */
+#define AUTOPAS_MPI_STATUS_IGNORE nullptr
+
+/**
+ * Dummy for MPI_Request.
+ */
+enum AutoPas_MPI_Request {
+  AUTOPAS_MPI_REQUEST_NULL,
+  _AUTOPAS_MPI_COMPLETED_REQUEST,
+  _AUTOPAS_MPI_INCOMPLETE_REQUEST,
```
Comment (SteffenSeckler) [inline src/autopas/utils/WrapMPI.h:125]: why did you add `_` here?

File: src/autopas/selectors/tuningStrategy/FullSearchMPI.h [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+      : _containerOptions(allowedContainerOptions),
+        _optimalConfig(Configuration()),
+        _configurationCommunicator(AutoPasConfigurationCommunicator()),
+        _localOptimalTime(0),
+        _request(AUTOPAS_MPI_REQUEST_NULL),
+        _allLocalConfigsTested(false),
+        _allGlobalConfigsTested(false),
+        _autopasMPICommunicator(comm) {
+    // @todo distribute the division process, so that not every rank has to traverse all configs
+    // @todo consider using MPI_Scatterv for dividing the search space
```
Comment (SteffenSeckler) [inline src/autopas/selectors/tuningStrategy/FullSearchMPI.h:50]: are these todos still valid?

File: src/autopas/AutoPas.h (hunk 4 above)
Comment (wothie) [reply src/autopas/AutoPas.h:122]: I'm only calling AutoPas_MPI_Comm_dup if has not been set until AutoPas::init() is called.
In finalize I don't know if the communicator has been created by the AutoPas_MPI_Comm_dup in init() or if it has been provided by the client. Thus if I free it blindly, the client implementation would crash if it also attempted to free it.
So either I have these rules or I explicitely store whether the communicator came externally.

File: src/autopas/utils/WrapMPI.h (hunk 8 above)
Comment (wothie) [reply src/autopas/utils/WrapMPI.h:125]: Because these do not correspond to anything defined by mpi. Ideally I would just want to not export them at all, but I chose this solution so that they at least wouldn't look like the regular dummy values.

File: src/autopas/selectors/tuningStrategy/FullSearchMPI.h (hunk 9 above)
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/FullSearchMPI.h:50]: > are these todos still valid?

The idea is still relevant, but considering that FullSearchMPI will be deleted soon, I suppose they aren't. I removed them.

**Comment by FG-TUM** [comment]
@wothie It seems the `AutoPasConfigurationCommunicator` is broken because we recently added a new element to the configurations. Can you please update it and add some instructions on what to do in `.github\CONTRIBUTING.md` maybe in a new subsection "Making an Option tuneable" (I will fill in the rest on what to do in such a case later.)?

**Comment by VincentF98** [comment]
@wothie My guide is up; see https://github.com/AutoPas/AutoPas/pull/487
